AdyenActions/Components/QRCode/QRCodeActionComponent.swift

```diff
@@ -0,0 +1,178 @@
+            )
+        }
+        
+        startTimer()
+    }
+    
+    /// :nodoc
+    private func startTimer() {
+        let unitCount = Int64(expirationTimeout)
+        progress.totalUnitCount = unitCount
+        progress.completedUnitCount = unitCount
+        
+        timeoutTimer = ExpirationTimer(
+            expirationTimeout: self.expirationTimeout,
+            onTick: { [weak self] in self?.updateExpiration($0) },
+            onExpiration: { [weak self] in self?.onTimerTimeout() }
+        )
+        timeoutTimer?.startTimer()
+    }
+    
+    private func updateExpiration(_ timeLeft: TimeInterval) {
+        progress.completedUnitCount = Int64(timeLeft)
+        let timeLeftString = timeLeft.adyen.timeLeftString() ?? ""
+        expirationText = localizedString(.pixExpirationLabel,
+                                         configuration.localizationParameters,
+                                         timeLeftString)
+    }
+    
+    private func onTimerTimeout() {
+        cleanup()
+        
+        delegate?.didFail(with: QRCodeComponentError.qrCodeExpired, from: self)
+    }
+    
+    private func createViewController(with action: QRCodeAction) -> UIViewController {
+        let viewController = QRCodeViewController(viewModel: createModel(with: action))
+        viewController.qrCodeView.delegate = self
+        return viewController
+    }
+    
+    private func createModel(with action: QRCodeAction) -> QRCodeView.Model {
+        let url = LogoURLProvider.logoURL(withName: action.paymentMethodType.rawValue, environment: context.apiContext.environment)
+        return QRCodeView.Model(
+            action: action,
+            instruction: localizedString(.pixInstructions,
+                                         configuration.localizationParameters),
+            payment: context.payment,
+            logoUrl: url,
+            observedProgress: progress,
+            expiration: $expirationText,
+            style: QRCodeView.Model.Style(
+                saveAsImageButton: configuration.style.saveAsImageButton,
+                instructionLabel: configuration.style.instructionLabel,
+                amountToPayLabel: configuration.style.amountToPayLabel,
+                progressView: configuration.style.progressView,
+                expirationLabel: configuration.style.expirationLabel,
+                logoCornerRounding: configuration.style.logoCornerRounding,
+                backgroundColor: configuration.style.backgroundColor
+            )
+        )
+    }
+    
+    public func didCancel() {
+        cleanup()
+    }
+    
+    fileprivate func cleanup() {
+        timeoutTimer?.stopTimer()
+    }
+
+}
+
+@_spi(AdyenInternal)
+extension QRCodeActionComponent: QRCodeViewDelegate {
+
+    internal func saveAsImage(qrCodeImage: UIImage?, sourceView: UIView) {
+        presentSharePopover(with: qrCodeImage as Any, sourceView: sourceView)
+    }
+    
+}
```

Comment:
this file is very similar to `PixComponent`, I am afraid there are quite a bit of code duplication here